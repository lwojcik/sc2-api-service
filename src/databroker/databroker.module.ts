import { Module } from '@nestjs/common';
import { Starcraft2ApiModule } from '../starcraft2api/starcraft2api.module';
import { CacheModule } from '../cache/cache.module';
import { LoggerModule } from '../logger/logger.module';
import { DataBrokerService } from './databroker.service';

@Module({
  imports: [CacheModule, Starcraft2ApiModule, LoggerModule],
  providers: [DataBrokerService],
  exports: [DataBrokerService],
})
export class DataBrokerModule {}
