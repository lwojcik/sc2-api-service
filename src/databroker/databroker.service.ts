import { Inject, Injectable } from '@nestjs/common';
import { ConfigType } from '@nestjs/config';
import { StarCraft2ApiService } from '../starcraft2api/starcraft2api.service';
import { CacheService } from '../cache/cache.service';
import { LoggerService } from '../logger/logger.service';
import { GetDataDto } from './dto/get-data.dto';
import { redisConfig } from '../config';
import { ApiData } from '../common/types';

@Injectable()
export class DataBrokerService {
  constructor(
    private readonly cacheService: CacheService,
    private readonly starCraft2ApiService: StarCraft2ApiService,
    @Inject(redisConfig.KEY)
    private readonly redisConf: ConfigType<typeof redisConfig>,
    private readonly logger: LoggerService
  ) {}

  private getDataKey(key: string, args: unknown) {
    const argString = Object.values(args).join(':');
    const dataKey = `:${key}:${argString}`;

    this.logger.debug(`Generated data key: ${dataKey}`);
    return dataKey;
  }

  private getDataFromCache(key: string) {
    return this.cacheService.get(key);
  }

  private cacheData(key: string, value: string) {
    return this.cacheService.set(key, value);
  }

  private async getDataAndRefreshCache(key: string, args: unknown) {
    this.logger.debug('Getting data from Battle.net...');

    const dataFromApi = await this.starCraft2ApiService.get(key, args);
    const dataKey = this.getDataKey(key, args);

    this.logger.debug(`Using data key: ${dataKey}`);

    if (
      dataFromApi.statusCode === 200 &&
      !((dataFromApi as ApiData).data as unknown as { error: string }).error
    ) {
      this.logger.debug('Caching data...');
      this.cacheData(
        dataKey,
        JSON.stringify({
          created: Math.floor(Date.now() / 1000),
          ...dataFromApi,
        })
      );
    }
    return dataFromApi;
  }

  async getData({ key, args }: GetDataDto, refresh = false) {
    const dataKey = this.getDataKey(key, args);
    this.logger.debug(`Using data key: ${dataKey}`);

    const cachedObject = await this.getDataFromCache(dataKey);

    if (!cachedObject || refresh) {
      this.logger.debug(
        'No cached data or refresh was triggered - getting fresh data...'
      );
      return this.getDataAndRefreshCache(key, args);
    }

    const parsedData = JSON.parse(cachedObject);
    const now = Math.floor(Date.now() / 1000);
    const staleData =
      now - Number(parsedData?.created) >= this.redisConf.ttlSecs;

    if (staleData) {
      this.logger.debug('Cached data is stale - getting fresh data...');
      this.getDataAndRefreshCache(key, args);
    }

    return cachedObject;
  }
}
