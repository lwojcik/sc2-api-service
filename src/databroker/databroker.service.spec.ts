import { Test, TestingModule } from '@nestjs/testing';
import { ModuleMocker } from 'jest-mock';
import { DataBrokerService } from './databroker.service';

const moduleMocker = new ModuleMocker(global);

describe('DataBrokerService', () => {
  let service: DataBrokerService;

  beforeEach(async () => {
    const module: TestingModule = await Test.createTestingModule({
      providers: [DataBrokerService],
    })
      .useMocker((token) => {
        if (typeof token === 'function') {
          const mockMetadata = moduleMocker.getMetadata(token);
          const Mock = moduleMocker.generateFromMetadata(mockMetadata);
          return new Mock();
        }
        return jest.fn();
      })
      .compile();

    service = module.get<DataBrokerService>(DataBrokerService);
  });

  it('should be defined', () => {
    expect(service).toBeDefined();
  });
});
