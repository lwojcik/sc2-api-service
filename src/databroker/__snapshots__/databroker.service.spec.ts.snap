// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DataBrokerService should get data by key 1`] = `
Object {
  "data": Object {
    "foo": "sample_data_from_sc2api_service_mock",
  },
  "statusCode": 200,
}
`;
